_posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md

```diff
@@ -0,0 +1,409 @@
+
+# Certificate Authorities and Let's Encrypt
+
+When you want to enable HTTPS on your website or need certificates for TLS communication, you'll need to request this certificate from a [Certificate Authority](https://en.wikipedia.org/wiki/Certificate_authority){:target="_blank" rel="noopener noreferrer"} (CA).
+It acts as a [trusted third party](https://en.wikipedia.org/wiki/Trusted_third_party){:target="_blank" rel="noopener noreferrer"} between two parties that need to communicate with each other.
+[Let's Encrypt](https://certbot.eff.org/){:target="_blank" rel="noopener noreferrer"} is such a Certificate Authority.
+It is their mission to give everyone a secure and privacy-respecting web experience.
+That's why they issue certificates free of charge.
+
+# Installing a Let's Encrypt certificate
+
+Assuming that you have shell access to your server, Let's Encrypt recommends to use [Certbot ACME Client](https://certbot.eff.org/){:target="_blank" rel="noopener noreferrer"}, since it can automate certificate issuance and installation with zero downtime.
+
+Certbot is a free, open source software tool for automatically using Let’s Encrypt certificates on manually-administrated websites to enable HTTPS.
+
+On the Certbot website you can find clear instructions on how to install.
+You can select your web server software (Apache, Nginx, ...) and operating system and Certbot provides the instructions to install it.
+
+> You can check your operating system on Linux by executing `cat /etc/os-release`.
+
+Please note that these instructions also include setting up HTTPS for your website, which for this tutorial isn't necessary.
+We'll use the certificate in another way, for TLS communication in a Java application.
+
+For Ubuntu, the following steps are required to install Certbot.
+See also [Apache on Ubuntu 16.04 (xenial)](https://certbot.eff.org/lets-encrypt/ubuntuxenial-apache){:target="_blank" rel="noopener noreferrer"}.
+
+Certbot is installed using APT (Advanced Package Tool), a tool for installing and removing applications on your system. This tool searches in its repositories for software distributions. Before you can install Certbot, you'll need to add the Certbot PPA (Personal Package Archive) to your list of available APT repositories.
+
+```
+sudo apt-get update
+sudo apt-get install software-properties-common
+sudo add-apt-repository universe
+sudo add-apt-repository ppa:certbot/certbot
+sudo apt-get update
+```
+
+Run the following command to install Certbot.
+
+```
+sudo apt-get install certbot python-certbot-apache
+```
+
+You can choose to retrieve a certificate, or retrieve it and install it immediately on your web server, eg. Apache by adding the parameter `--apache`.
+For our situation, it is enough to retrieve a certificate.
+You do this with the following command:
+
+```
+certbot certonly
+```
+
+You can find the installed Let's Encrypt certificates in the `/etc/letsencrypt/live` folder on your file system.
+
+# Certificate renewal
```

Comment:
Certbot [automatically renews](https://certbot.eff.org/docs/using.html?highlight=hooks#automated-renewals) certificates on most operating systems now.
you can verify this by looking at `systemctl list-timers`.

```
NEXT                         LEFT        LAST                         PASSED              UNIT                         ACTIVATES
Wed 2019-08-14 10:37:47 UTC  15h left    n/a                          n/a                 certbot.timer                certbot.service
```

Testing the renewal is a good practice though. Example output of `certbot renew --dry-run`:
```
Saving debug log to /var/log/letsencrypt/letsencrypt.log

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Processing /etc/letsencrypt/renewal/mydomain.be.conf
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Cert not due for renewal, but simulating renewal for dry run
Plugins selected: Authenticator standalone, Installer None
Renewing an existing certificate

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
new certificate deployed without reload, fullchain is
/etc/letsencrypt/live/mydomain.be/fullchain.pem
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
** DRY RUN: simulating 'certbot renew' close to cert expiry
**          (The test certificates below have not been saved.)

Congratulations, all renewals succeeded. The following certs have been renewed:
  /etc/letsencrypt/live/mydomain.be/fullchain.pem (success)
** DRY RUN: simulating 'certbot renew' close to cert expiry
**          (The test certificates above have not been saved.)
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
```

Reply:
Nice catch, thanks! 2y ago that wasn't the case.
I'll rewrite this part and also mention the automatic cron setup.